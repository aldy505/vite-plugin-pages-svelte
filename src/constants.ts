export const MODULE_IDS = ['pages-generated-svelte', 'virtual:generated-pages-svelte'];
export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages-svelte';
