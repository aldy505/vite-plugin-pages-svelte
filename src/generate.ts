import type { PreRoute } from './types/route';
import type { FileOutput } from './types/page';
import { isDynamicRoute, isCatchAllRoute } from './utils/validate';
import { stringifyRoutes } from './stringify';
import { haveChildren } from './crawler/crawler';
import { extname } from 'path';

export function generateRoutes(pages: FileOutput[]): PreRoute[] {
  const routes: PreRoute[] = [];

  for (let i = 0; i < pages.length; i++) {
    const node = pages[i].path.split('/')[pages[i].path.split('/').length - 1];
    const isDynamic = isDynamicRoute(node);
    const isCatchAll = isCatchAllRoute(node);
    const fileExt = extname(node);
    const normalizedName = isDynamic
      ? node
          .replace(/^\[(\.{3})?/, '')
          .replace(/\]$/, '')
          .replace(fileExt, '')
      : node.replace(fileExt, '');
    const normalizedPath = normalizedName.toLowerCase();
    let name: string;
    if (normalizedPath === 'index') {
      name = 'index';
    } else {
      if (isCatchAll) {
        name = '/*';
      } else if (isDynamic) {
        name = `/:${normalizedName}`;
      } else {
        name = `/${normalizedPath}`;
      }
    }

    if (!haveChildren(pages[i])) {
      routes.push({
        name,
        path: pages[i].path,
      });
      continue;
    }

    routes.push({
      name,
      children: generateRoutes(pages[i].children as FileOutput[]),
    });
  }

  return routes;
}

export function generateClientCode(routes: PreRoute[]): string {
  const { imports, stringRoutes } = stringifyRoutes(routes);

  return (
    `${imports.join(';\n')}${imports.length > 1 ? ';' : ''}\n\n` +
    `const routes = ${stringRoutes};\n\n` +
    `export default routes;`
  );
}
