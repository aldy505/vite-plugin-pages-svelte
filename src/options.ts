import type { UserOptions, ResolvedOptions } from './types/options';
import { slash } from './utils/convert';

/**
 *
 * @param userOptions
 * @param viteRoot
 * @returns
 */
export async function resolveOptions(userOptions: UserOptions, viteRoot?: string): Promise<ResolvedOptions> {
  const { pagesDir = 'src/pages', exclude = [], syncIndex = true } = userOptions;

  const root = viteRoot || slash(process.cwd());

  const extensions = userOptions.extensions || ['svelte'];

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`);

  const resolvedOptions: ResolvedOptions = {
    pagesDir,
    root,
    extensions,
    exclude,
    syncIndex,
    extensionsRE,
  };

  return resolvedOptions;
}
