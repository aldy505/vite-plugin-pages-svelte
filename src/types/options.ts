/**
 * Plugin options.
 */
interface Options {
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string;
  /**
   * Valid file extensions for page components.
   * @default ['svelte']
   */
  extensions: string[];
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[];
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean;
}

export type UserOptions = Partial<Options>;

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string;
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp;
}

export type ImportMode = 'sync' | 'async';
export type ImportModeResolveFn = (filepath: string) => ImportMode;
