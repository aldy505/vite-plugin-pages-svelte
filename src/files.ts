import fg from 'fast-glob';
import { PageDirOptions, ResolvedOptions } from './types';
import { extensionsToGlob } from './utils';

function getIgnore(exclude: string[]) {
  return ['node_modules', '.git', '**/__*__/**', ...exclude];
}

/**
 * Resolves the page dirs for its for its given globs
 */
export function getPageDirs(pageDirOptions: PageDirOptions, root: string, exclude: string[]): PageDirOptions[] {
  const dirs = fg.sync(pageDirOptions.dir, {
    ignore: getIgnore(exclude),
    onlyDirectories: true,
    dot: true,
    unique: true,
    cwd: root,
  });

  const pageDirs = dirs.map((dir) => ({
    ...pageDirOptions,
    dir,
  }));

  return pageDirs;
}

/**
 * Resolves the files that are valid pages for the given context.
 */
export function getPageFiles(path: string, options: ResolvedOptions): string[] {
  const { exclude, extensions } = options;

  const ext = extensionsToGlob(extensions);

  const files = fg.sync(`**/*.${ext}`, {
    ignore: getIgnore(exclude),
    onlyFiles: true,
    cwd: path,
  });

  return files;
}
