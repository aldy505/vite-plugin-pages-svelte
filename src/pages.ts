import type { ResolvedOptions } from './types/options';
import type { FileOutput } from './types/page';
import { resolve } from 'path';
import { getPageFiles } from './files';
import { slash } from './utils/convert';

export async function resolvePages(options: ResolvedOptions): Promise<FileOutput[]> {
  const pages: FileOutput[] = [];

  const pageDirFiles = {
    pageDir: options.pagesDir,
    files: await getPageFiles(slash(resolve(options.root, options.pagesDir)), options),
  };

  for (const file of pageDirFiles.files) {
    pages.push(file);
  }

  const routes: string[] = [];

  for (const page of pages.values()) {
    if (!routes.includes(page.path)) routes.push(page.path);
    else throw new Error(`[vite-plugin-pages] duplicate route in ${page.path}`);
  }

  return pages;
}

export function addPage(pages: FileOutput[], newPage: FileOutput): void {
  pages.push(newPage);
}

export function removePage(pages: FileOutput[], oldPage: FileOutput): void {
  const find = pages.findIndex((o) => o.path === oldPage.path);
  pages.splice(find, 1);
}
