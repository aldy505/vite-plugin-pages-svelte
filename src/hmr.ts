import { ViteDevServer } from 'vite';
import { getPagesVirtualModule, isRouteBlockChanged, isTarget, debug, slash } from './utils';
import { removePage, addPage, updatePage } from './pages';
import { ResolvedOptions, ResolvedPages } from './types';

export function handleHMR(
  server: ViteDevServer,
  pages: ResolvedPages,
  options: ResolvedOptions,
  clearRoutes: () => void,
): void {
  const { ws, watcher } = server;

  function fullReload() {
    // invalidate module
    getPagesVirtualModule(server);
    clearRoutes();
    ws.send({
      type: 'full-reload',
    });
  }

  watcher.on('add', async (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      await addPage(pages, path, options);
      debug.hmr('add', path);
      fullReload();
    }
  });
  watcher.on('unlink', (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      removePage(pages, path);
      debug.hmr('remove', path);
      fullReload();
    }
  });
  watcher.on('change', async (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      const needReload = await isRouteBlockChanged(path, options);
      if (needReload) {
        updatePage(pages, path);
        debug.hmr('change', path);
        fullReload();
      }
    }
  });
}
