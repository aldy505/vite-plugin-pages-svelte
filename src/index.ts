import type { Plugin } from 'vite';
import type { ResolvedOptions, UserOptions } from './types/options';
import type { PreRoute } from './types/route';
import type { FileOutput } from './types/page';
import { generateRoutes, generateClientCode } from './generate';
import { debug } from './utils/vite';
import { resolveOptions } from './options';
import { MODULE_IDS, MODULE_ID_VIRTUAL } from './constants';
import { resolvePages } from './pages';
import { handleHMR } from './hmr';

function pagesPlugin(userOptions: UserOptions = {}): Plugin {
  let generatedRoutes: PreRoute[] | null = null;
  let options: ResolvedOptions;
  let pages: FileOutput[];

  return {
    name: 'vite-plugin-pages-svelte',
    enforce: 'pre',
    async configResolved({ root }) {
      options = await resolveOptions(userOptions, root);
      pages = await resolvePages(options);
      debug.options(options);
      debug.pages(pages);
    },
    configureServer(server) {
      handleHMR(server, pages, options, () => {
        generatedRoutes = null;
      });
    },
    resolveId(id) {
      return MODULE_IDS.includes(id) || MODULE_IDS.some((i) => id.startsWith(i)) ? MODULE_ID_VIRTUAL : null;
    },
    load(id) {
      if (id !== MODULE_ID_VIRTUAL) return;

      if (!generatedRoutes) {
        generatedRoutes = generateRoutes(pages);
      }

      const clientCode = generateClientCode(generatedRoutes);
      debug.gen('client code: %O', clientCode);

      return clientCode;
    },
    transform(_code, id) {
      if (!/svelte&type=route/.test(id)) return;
      return {
        code: 'export default {};',
        map: null,
      };
    },
    generateBundle(_options, bundle) {
      bundle;
    },
  };
}

export * from './types/options';
export { generateRoutes };
export default pagesPlugin;
