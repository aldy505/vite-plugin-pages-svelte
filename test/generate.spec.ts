import { generateRoutes, generateClientCode } from '../src/generate';
import { resolvePages } from '../src/pages';
import { resolveOptions } from '../src/options';

describe('Generate', () => {
  test('Routes Sync', async () => {
    const options = await resolveOptions({
      pagesDir: 'test/assets/pages',
    });
    const pages = await resolvePages(options);
    const routes = generateRoutes(pages);
    const code = generateClientCode(routes);

    expect(routes).toMatchSnapshot('routes sync');
    expect(code).toMatchSnapshot('client code sync');
  });
});
