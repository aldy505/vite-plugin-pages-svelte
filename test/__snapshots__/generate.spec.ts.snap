// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Routes Sync: client code sync 1`] = `
"import _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_about_svelte from \\"/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/about.svelte\\";
import _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_index_svelte from \\"/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/index.svelte\\";
import _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_$userId$_svelte from \\"/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[userId].svelte\\";
import _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_components_svelte from \\"/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/components.svelte\\";
import _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_$___all$_svelte from \\"/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[...all].svelte\\";

const routes = [{name: \\"/about\\", component: _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_about_svelte,},{name: \\"/about\\", nestedRoutes: [object Object]},{name: \\"index\\", component: _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_index_svelte,},{name: \\"/[userid]\\", component: _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_$userId$_svelte,},{name: \\"/components\\", component: _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_components_svelte,},{name: \\"/:sensor\\", nestedRoutes: [object Object][object Object]},{name: \\"/blog\\", nestedRoutes: [object Object][object Object][object Object]},{name: \\"/*\\", component: _home_reinaldy_repository_vite_plugin_pages_svelte_test_assets_pages_$___all$_svelte,},{name: \\"/__test__\\", nestedRoutes: [object Object]},];

export default routes;"
`;

exports[`Generate Routes Sync: routes sync 1`] = `
Array [
  Object {
    "name": "/about",
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/about.svelte",
  },
  Object {
    "children": Array [
      Object {
        "name": "index",
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/about/index.svelte",
      },
    ],
    "name": "/about",
  },
  Object {
    "name": "index",
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/index.svelte",
  },
  Object {
    "name": "/[userid]",
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[userId].svelte",
  },
  Object {
    "name": "/components",
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/components.svelte",
  },
  Object {
    "children": Array [
      Object {
        "name": "/current",
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[sensor]/current.svelte",
      },
      Object {
        "name": "/*",
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[sensor]/[...all].svelte",
      },
    ],
    "name": "/:sensor",
  },
  Object {
    "children": Array [
      Object {
        "name": "index",
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/index.svelte",
      },
      Object {
        "name": "/[id]",
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/[id].svelte",
      },
      Object {
        "children": Array [
          Object {
            "name": "index",
            "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/today/index.svelte",
          },
        ],
        "name": "/today",
      },
    ],
    "name": "/blog",
  },
  Object {
    "name": "/*",
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[...all].svelte",
  },
  Object {
    "children": Array [
      Object {
        "name": "index",
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/__test__/index.svelte",
      },
    ],
    "name": "/__test__",
  },
]
`;
